src/language/branches.md

```diff
@@ -0,0 +1,295 @@
+
+### Switch statements
+
+Non-empty `case` clauses implicitly jump to the end of the switch after completion.
+Other valid ways to end a non-empty `case` clause are a [`continue`][break],
+[`throw`][], or [`return`][] statement.
+
+Use a `default` or wildcard `_` clause to execute code when no `case` clause matches:
+
+```dart
+var command = 'OPEN';
+switch (command) {
+  case 'CLOSED':
+    executeClosed();
+  case 'PENDING':
+    executePending();
+  case 'APPROVED':
+    executeApproved();
+  case 'DENIED':
+    executeDenied();
+  case 'OPEN':
+    executeOpen();
+  default:
+    executeUnknown();
+}
+```
+
+Empty cases fall through to the next case. For an empty case that does not fall
+through, use [`break`][break] for its body:
+
+```dart
+var command = 'CLOSED';
+switch (command) {
+  case 'CLOSED': // Empty case falls through.
+  case 'NOW_CLOSED':
+    // Runs for both CLOSED and NOW_CLOSED.
+    executeNowClosed();
+}
+```
+
+For non-sequential fall-through, you can use a [`continue` statement][break] and
+a label:
+
+```dart
+var command = 'CLOSED';
+switch (command) {
+  case 'CLOSED':
+    executeClosed();
+    continue nowClosed;
+  // Continues executing at the nowClosed label.
+
+  nowClosed:
+  case 'NOW_CLOSED':
+    // Runs for both CLOSED and NOW_CLOSED.
+    executeNowClosed();
+}
+```
+
+### Switch expressions
+
+_Switch expressions_ write a result value determined by a switch
+directly into a variable or return statement. 
+
+The syntax rules for a switch expression are:
+- The expression body is a series of case patterns.
+- Cases _do not_ use the `case` keyword.
+- Case patterns are followed by `=>`, and then a case body. 
+- Case bodies terminate with a comma `,`.
+- The switch expression body terminates with a semicolon `;`.
+- The wildcard pattern `_` catches remaining values instead of `default`.
+
+This allows you to rewrite a 
+_statement_ like this:
+
+```dart
+switch (charCode) {
+  case slash when nextCharCode == slash:   // Constant pattern with guard
+    skipComment();
+  case slash || star || plus || minus:     // Logical-or pattern
+    operator(charCode);
+  case >= digit0 && <= digit9:             // Relational and logical-and patterns
+    number();
+  default:
+    invalid();
+}
+```
+
+Into an _expression_, like this:
+
+```dart
+var token = switch (charCode) {
+  slash when nextCharCode == slash => skipComment(),
+  slash || star || plus || minus   => operator(charCode),
+  >= digit0 && <= digit9           => number(),
+  _                                => throw invalid()
+};
+```
+
+### Exhaustiveness checking
+
+Exhaustiveness checking is a feature of switches that raises a compile-time error
+if a switch is not exhaustively covering all possible values that could flow through it.
+
+The compiler can only know if a set of cases is
+exhaustive if it is soundly aware of all possible enumerations of that type. So,
+exhaustiveness checking works only when switching on types that are fully enumerable:
```

Comment:
Exhaustiveness does actually work for all kinds of types. (Otherwise switch expressions would be much less useful because they *have* to be exhaustive. Otherwise, you might not know what value the expression evaluates to.)

The reason that enums and sealed types are particularly useful for switches is that they allow you to "partition" the matched value in a way such that multiple cases each of which handle only a subset of the possible values can provably cover all possible values.

For other types, to be exhaustive, you basically need to have a single case that covers everything. So in:

```dart
int n = ...
var name = switch (n) {
  1 => 'one',
  2 => 'two',
  _ => 'other'
};
```

The switch is exhaustive because the final `_` pattern covers all possible ints.

I'd probably just remove this section. I don't think we go into a lot of detail about how type inference and flow analysis work and those are similar where there's a decent amount of algorithm sophistication going on under the hood and we just expect that it aligns with user intuition.

So if they have an intuition about the set of all possible values going into a switch and a good intuition about which values the set of patterns covers then exhaustiveness checking should mostly align with that. The actual mechanics of it get pretty complex.